01-Getting-Started.md

```diff
@@ -160,3 +173,129 @@ Vanagon contains the `build` command for generating a package. It also contains
+cd cpp-project-template/build
+cmake -G "MinGW Makefiles" -DBOOST_STATIC=ON -DBOOST_ROOT="${toolsDir}\boost_1_58_0-x86_64_mingw-w64_4.8.3_win32_seh" -DCURL_STATIC=ON -DCMAKE_PREFIX_PATH="${toolsDir}\curl-7.42.1-x86_64_mingw-w64_4.8.3_win32_seh.7z" ..
+mingw32-make -j
+```
+
+TODO: Host Boost and other libraries on myget.org to install with chocolatey, and simplify CMAKE_PREFIX_PATH. Also use curl with openssl.
+
+##### Testing
+
+To run tests with verbose output, run `make test ARGS=-V` (or `mingw32-make test ARGS=-V` on Windows).
+
+#### File Paths and System Calls
+
+Obviously file paths on platforms differ. Refer to the [file_paths specification](https://github.com/puppetlabs/puppet-specifications/blob/master/file_paths.md).
+
+Assume POSIX functions aren't available on Windows, and prefer Boost or C++ standard library functions. If specific POSIX functionality is needed, an equivalent will be needed for Windows; such helpers are good candidates for a library in Leatherman.
+
+TODO: Add ways to get common paths on Windows.
+
+#### Unicode and UTF-8
+
+Windows chose UTF-16 as the default Unicode representation, and all system calls are based on that; C++ standard library functions only support ASCII or other fixed 8-bit encodings (such as [Latin-1](https://en.wikipedia.org/wiki/Windows-1252)). UNIX-based platforms generally support UTF-8 in system calls and the C++ standard library. `std::string` is just a byte container, so it safely holds UTF-8 characters; however when manipulating UTF-8 strings you should use a UTF-8 aware library such as [UTF8-CPP](http://utfcpp.sourceforge.net/) or Boost.Regex. Our practice is to represent strings as UTF-8 internally, and deal with conversion when text is read/written.
+
+TODO: Evaluate UTF8-CPP and put it on Library list.
+
+##### Handling Input/Output
+
+The [driver in cpp-project-template](https://github.com/puppetlabs/cpp-project-template/blob/master/exe/driver.cc.in) contains useful examples of handling input/output safely across platforms.
+
+```
+// Use Boost.Nowide args to ensure command-line arguments are UTF-8.
+// On Windows this accesses console arguments as UTF-16 and converts them to UTF-8.
+// This uses the RAII pattern to modify argc/argv in-place, and revert them when exiting scope.
+boost::nowide::args arg_utf8(argc, argv);
+
+// Use Boost.Nowide cout/cerr to correctly print UTF-8 characters.
+// On Windows this converts to UTF-16 before printing to the console.
+setup_logging(boost::nowide::cerr);
+boost::nowide::cout << "Hello!" << std::endl;
+
+// Use Boost.Nowide cin to correctly read UTF-8 characters.
+// On Windows this reads using a UTF-16 system call and converts to UTF-8.
+std::string input;
+boost::nowide::cin >> input;
+```
+
+Boost.Nowide's cout/cerr implementations come with the caveat that they aren't header-only, so if statically linking against Boost.Nowide flushing in different libraries will not be synchronized or affect text written in other libraries.
```

Comment:
I see, perhaps `...so if Boost.Nowide is statically linked into multiple shared libraries, input and output must be manually synchronized between those libraries`?